# Data Understanding

The goal of this notebook is to inspect the dataset before performing any preprocessing or model training.

In [12]:
import pandas as pd

pd.set_option("display.max_columns", None)

df = pd.read_excel("../data/raw/telco_customer_churn.xlsx")
df.head()

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,Senior Citizen,Partner,Dependents,Tenure Months,Phone Service,Multiple Lines,Internet Service,Online Security,Online Backup,Device Protection,Tech Support,Streaming TV,Streaming Movies,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
0,3668-QPYBK,1,United States,California,Los Angeles,90003,"33.964131, -118.272783",33.964131,-118.272783,Male,No,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes,1,86,3239,Competitor made better offer
1,9237-HQITU,1,United States,California,Los Angeles,90005,"34.059281, -118.30742",34.059281,-118.307420,Female,No,No,Yes,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes,1,67,2701,Moved
2,9305-CDSKC,1,United States,California,Los Angeles,90006,"34.048013, -118.293953",34.048013,-118.293953,Female,No,No,Yes,8,Yes,Yes,Fiber optic,No,No,Yes,No,Yes,Yes,Month-to-month,Yes,Electronic check,99.65,820.5,Yes,1,86,5372,Moved
3,7892-POOKP,1,United States,California,Los Angeles,90010,"34.062125, -118.315709",34.062125,-118.315709,Female,No,Yes,Yes,28,Yes,Yes,Fiber optic,No,No,Yes,Yes,Yes,Yes,Month-to-month,Yes,Electronic check,104.80,3046.05,Yes,1,84,5003,Moved
4,0280-XJGEX,1,United States,California,Los Angeles,90015,"34.039224, -118.266293",34.039224,-118.266293,Male,No,No,Yes,49,Yes,Yes,Fiber optic,No,Yes,Yes,No,Yes,Yes,Month-to-month,Yes,Bank transfer (automatic),103.70,5036.3,Yes,1,89,5340,Competitor had better devices


## Dimensions

In [13]:
df.shape

(7043, 33)

## Data Types

In [14]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   str    
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   str    
 3   State              7043 non-null   str    
 4   City               7043 non-null   str    
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   str    
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   str    
 10  Senior Citizen     7043 non-null   str    
 11  Partner            7043 non-null   str    
 12  Dependents         7043 non-null   str    
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   str    
 15  Multiple Lines     7043 non-null   str    
 16  Internet Service   7043 non-null   

### Investigation of the Total Charges feature

In [15]:
print(df["Total Charges"].value_counts())

Total Charges
20.2      11
          11
19.75      9
19.9       8
19.65      8
          ..
1419.4     1
1990.5     1
7362.9     1
346.45     1
6844.5     1
Name: count, Length: 6531, dtype: int64


In [16]:
(df["Total Charges"].str.strip() == "").sum()

np.int64(11)

In [17]:
df[df["Total Charges"].str.strip() == ""]

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,Senior Citizen,Partner,Dependents,Tenure Months,Phone Service,Multiple Lines,Internet Service,Online Security,Online Backup,Device Protection,Tech Support,Streaming TV,Streaming Movies,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
2234,4472-LVYGI,1,United States,California,San Bernardino,92408,"34.084909, -117.258107",34.084909,-117.258107,Female,No,Yes,No,0,No,No phone service,DSL,Yes,No,Yes,Yes,Yes,No,Two year,Yes,Bank transfer (automatic),52.55,,No,0,36,2578,NaN
2438,3115-CZMZD,1,United States,California,Independence,93526,"36.869584, -118.189241",36.869584,-118.189241,Male,No,No,No,0,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.25,,No,0,68,5504,NaN
2568,5709-LVOEQ,1,United States,California,San Mateo,94401,"37.590421, -122.306467",37.590421,-122.306467,Female,No,Yes,No,0,Yes,No,DSL,Yes,Yes,Yes,No,Yes,Yes,Two year,No,Mailed check,80.85,,No,0,45,2048,NaN
2667,4367-NUYAO,1,United States,California,Cupertino,95014,"37.306612, -122.080621",37.306612,-122.080621,Male,No,Yes,Yes,0,Yes,Yes,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.75,,No,0,48,4950,NaN
2856,1371-DWPAZ,1,United States,California,Redcrest,95569,"40.363446, -123.835041",40.363446,-123.835041,Female,No,Yes,No,0,No,No phone service,DSL,Yes,Yes,Yes,Yes,Yes,No,Two year,No,Credit card (automatic),56.05,,No,0,30,4740,NaN
4331,7644-OMVMY,1,United States,California,Los Angeles,90029,"34.089953, -118.294824",34.089953,-118.294824,Male,No,Yes,Yes,0,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,19.85,,No,0,53,2019,NaN
4687,3213-VVOLG,1,United States,California,Sun City,92585,"33.739412, -117.173334",33.739412,-117.173334,Male,No,Yes,Yes,0,Yes,Yes,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.35,,No,0,49,2299,NaN
5104,2520-SGTTA,1,United States,California,Ben Lomond,95005,"37.078873, -122.090386",37.078873,-122.090386,Female,No,Yes,Yes,0,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.00,,No,0,27,3763,NaN
5719,2923-ARZLG,1,United States,California,La Verne,91750,"34.144703, -117.770299",34.144703,-117.770299,Male,No,Yes,Yes,0,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,One year,Yes,Mailed check,19.70,,No,0,69,4890,NaN
6772,4075-WKNIU,1,United States,California,Bell,90201,"33.970343, -118.171368",33.970343,-118.171368,Female,No,Yes,Yes,0,Yes,Yes,DSL,No,Yes,Yes,Yes,Yes,No,Two year,No,Mailed check,73.35,,No,0,44,2342,NaN


In [18]:
df[df["Total Charges"].str.strip() == ""].describe(include="all")

,CustomerID,Count,Country,State,City,Zip Code,Lat Long,Latitude,Longitude,Gender,Senior Citizen,Partner,Dependents,Tenure Months,Phone Service,Multiple Lines,Internet Service,Online Security,Online Backup,Device Protection,Tech Support,Streaming TV,Streaming Movies,Contract,Paperless Billing,Payment Method,Monthly Charges,Total Charges,Churn Label,Churn Value,Churn Score,CLTV,Churn Reason
count,11,11.0,11,11,11,11.000000,11,11.000000,11.000000,11,11,11,11,11.0,11,11,11,11,11,11,11,11,11,11,11,11,11.000000,11,11,11.0,11.000000,11.000000,0
unique,11,NaN,1,1,11,NaN,11,NaN,NaN,2,1,2,2,NaN,2,3,2,3,3,3,3,3,3,2,2,3,NaN,1,1,NaN,NaN,NaN,0
top,4472-LVYGI,NaN,United States,California,San Bernardino,NaN,"34.084909, -117.258107",NaN,NaN,Male,No,Yes,Yes,NaN,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,NaN,,No,NaN,NaN,NaN,NaN
freq,1,NaN,11,11,1,NaN,1,NaN,NaN,6,11,9,7,NaN,9,5,6,6,6,6,6,6,6,10,8,8,NaN,11,11,NaN,NaN,NaN,NaN
mean,NaN,1.0,NaN,NaN,NaN,92839.272727,NaN,35.729120,-119.584723,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,41.418182,NaN,NaN,0.0,48.545455,3665.545455,NaN
std,NaN,0.0,NaN,NaN,NaN,2011.647588,NaN,2.220277,2.446604,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,23.831484,NaN,NaN,0.0,14.445509,1419.168233,NaN
min,NaN,1.0,NaN,NaN,NaN,90029.000000,NaN,33.739412,-123.835041,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,19.700000,NaN,NaN,0.0,27.000000,2019.000000,NaN
25%,NaN,1.0,NaN,NaN,NaN,91247.000000,NaN,34.027626,-122.085503,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,20.125000,NaN,NaN,0.0,40.000000,2320.500000,NaN
50%,NaN,1.0,NaN,NaN,NaN,92585.000000,NaN,34.144703,-118.262263,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,25.750000,NaN,NaN,0.0,48.000000,3763.000000,NaN
75%,NaN,1.0,NaN,NaN,NaN,94703.000000,NaN,37.192743,-117.970833,NaN,NaN,NaN,NaN,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,58.975000,NaN,NaN,0.0,59.000000,4920.000000,NaN


In [20]:
df.loc[
    df["Total Charges"].str.strip() == "",
    ["Tenure Months", "Monthly Charges", "Total Charges", "Churn Label"]
]

,Tenure Months,Monthly Charges,Total Charges,Churn Label
2234,0,52.55,,No
2438,0,20.25,,No
2568,0,80.85,,No
2667,0,25.75,,No
2856,0,56.05,,No
4331,0,19.85,,No
4687,0,25.35,,No
5104,0,20.00,,No
5719,0,19.70,,No
6772,0,73.35,,No


During the initial inspection, we identified that the Total Charges feature is stored as an object instead of a numeric data type. Further investigation revealed that this is caused by 11 empty string values. These records correspond to customers with zero tenure, suggesting that the missing total charges may represent expected business behavior rather than erroneous data. The column will be handled during preprocessing.

## Missing Values

In [21]:
df.isna().sum()

CustomerID              0
Count                   0
Country                 0
State                   0
City                    0
Zip Code                0
Lat Long                0
Latitude                0
Longitude               0
Gender                  0
Senior Citizen          0
Partner                 0
Dependents              0
Tenure Months           0
Phone Service           0
Multiple Lines          0
Internet Service        0
Online Security         0
Online Backup           0
Device Protection       0
Tech Support            0
Streaming TV            0
Streaming Movies        0
Contract                0
Paperless Billing       0
Payment Method          0
Monthly Charges         0
Total Charges           0
Churn Label             0
Churn Value             0
Churn Score             0
CLTV                    0
Churn Reason         5174
dtype: int64

Missing values were detected only in the Churn Reason feature. Since this information is available only for customers who have churned, the missing values are expected rather than indicating poor data quality.

The Churn Reason feature records the reason why a customer left the company. Since this information becomes available only after the churn event has occurred, it would not be available at prediction time. Using this feature during model training would introduce data leakage, resulting in unrealistically optimistic model performance.

## Duplicate Rows

In [22]:
df.duplicated().sum()

np.int64(0)

No duplicated observations were detected.

## Cardinality

In [23]:
df.nunique()

CustomerID           7043
Count                   1
Country                 1
State                   1
City                 1129
Zip Code             1652
Lat Long             1652
Latitude             1652
Longitude            1651
Gender                  2
Senior Citizen          2
Partner                 2
Dependents              2
Tenure Months          73
Phone Service           2
Multiple Lines          3
Internet Service        3
Online Security         3
Online Backup           3
Device Protection       3
Tech Support            3
Streaming TV            3
Streaming Movies        3
Contract                3
Paperless Billing       2
Payment Method          4
Monthly Charges      1585
Total Charges        6531
Churn Label             2
Churn Value             2
Churn Score            85
CLTV                 3438
Churn Reason           20
dtype: int64

Features with a cardinality of one do not contain any variability and therefore provide no discriminative information for the model. These features will be evaluated for removal during preprocessing.

## Numeric Statistics

In [24]:
df.describe()

,Count,Zip Code,Latitude,Longitude,Tenure Months,Monthly Charges,Churn Value,Churn Score,CLTV
count,7043.0,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000
mean,1.0,93521.964646,36.282441,-119.798880,32.371149,64.761692,0.265370,58.699418,4400.295755
std,0.0,1865.794555,2.455723,2.157889,24.559481,30.090047,0.441561,21.525131,1183.057152
min,1.0,90001.000000,32.555828,-124.301372,0.000000,18.250000,0.000000,5.000000,2003.000000
25%,1.0,92102.000000,34.030915,-121.815412,9.000000,35.500000,0.000000,40.000000,3469.000000
50%,1.0,93552.000000,36.391777,-119.730885,29.000000,70.350000,0.000000,61.000000,4527.000000
75%,1.0,95351.000000,38.224869,-118.043237,55.000000,89.850000,1.000000,75.000000,5380.500000
max,1.0,96161.000000,41.962127,-114.192901,72.000000,118.750000,1.000000,100.000000,6500.000000


The numerical features appear to contain valid values within reasonable ranges. No obvious inconsistencies, such as negative charges or unrealistic tenure values, were identified during the initial inspection. More analysis of the distributions and potential outliers will be performed during the exploratory data analysis stage.

## Categorical Statistics

In [26]:
df.describe(include="string")

,CustomerID,Country,State,City,Lat Long,Gender,Senior Citizen,Partner,Dependents,Phone Service,Multiple Lines,Internet Service,Online Security,Online Backup,Device Protection,Tech Support,Streaming TV,Streaming Movies,Contract,Paperless Billing,Payment Method,Churn Label,Churn Reason
count,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,1869
unique,7043,1,1,1129,1652,2,2,2,2,2,3,3,3,3,3,3,3,3,3,2,4,2,20
top,3668-QPYBK,United States,California,Los Angeles,"33.964131, -118.272783",Male,No,No,No,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,No,Attitude of support person
freq,1,7043,7043,305,5,3555,5901,3641,5416,6361,3390,3096,3498,3088,3095,3473,2810,2785,3875,4171,2365,5174,192


The categorical features contain plausible values and no unexpected categories were identified during the initial inspection. Several features, such as Country and State, contain only a single unique value and are therefore unlikely to provide predictive information.

# Key Findings

The initial inspection of the dataset revealed several important observations:

- The dataset contains 7,043 customer records and 33 features.
- The `Total Charges` feature is stored as an object due to 11 empty string values associated with customers having zero tenure.
- Missing values occur only in the `Churn Reason` feature and appear to be expected.
- Several features (`Count`, `Country`, and `State`) have no variability and will be considered for removal during preprocessing.
- The remaining numerical and categorical features appear consistent based on the initial inspection.

## Feature Assessment

| Feature | Role | Recommendation |
|---------|------|--------------------|
| CustomerID | Identifier | Remove before training |
| Count | Constant feature | Candidate for removal |
| Country | Constant feature | Candidate for removal |
| State | Constant feature | Candidate for removal |
| Total Charges | Numerical feature | Convert to numeric during preprocessing |
| Churn Label | Target variable | Prediction target |
| Churn Value | Encoded target | Redundant with `Churn Label` |
| Churn Score | Unknown | Investigate before modeling |
| CLTV | Business metric | Investigate before modeling |
| Churn Reason | Post-event information | Exclude to prevent data leakage |